pkg/controller/registry/resolver/cache.go

```diff
@@ -0,0 +1,287 @@
+	"github.com/operator-framework/operator-registry/pkg/client"
+)
+
+type ClientProvider interface {
+	Get() (client.Interface, error)
+}
+
+type LazyClient struct {
+	address string
+	c       client.Interface
+	m       sync.Mutex
+}
+
+func (lc *LazyClient) Get() (client.Interface, error) {
+	lc.m.Lock()
+	defer lc.m.Unlock()
+	if lc.c != nil {
+		return lc.c, nil
+	}
+	client, err := client.NewClient(lc.address)
+	if err != nil {
+		return nil, err
+	}
+	lc.c = client
+	return lc.c, nil
+}
+
+type RegistryClientProvider interface {
+	ClientsForNamespaces(namespaces ...string) map[CatalogKey]ClientProvider
+}
+
+type DefaultRegistryClientProvider struct {
+	logger logrus.FieldLogger
+	c      versioned.Interface
+}
+
+func NewDefaultRegistryClientProvider(c versioned.Interface) *DefaultRegistryClientProvider {
+	return &DefaultRegistryClientProvider{
+		logger: logrus.New(),
+		c:      c,
+	}
+}
+
+func (rcp *DefaultRegistryClientProvider) ClientsForNamespaces(namespaces ...string) map[CatalogKey]ClientProvider {
+	result := make(map[CatalogKey]ClientProvider)
+	for _, namespace := range namespaces {
+		list, err := rcp.c.OperatorsV1alpha1().CatalogSources(namespace).List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			rcp.logger.Errorf("failed to list catalogsources in %s: %s", namespace, err.Error())
+			continue
+		}
+		for _, source := range list.Items {
+			if source.Status.RegistryServiceStatus == nil {
+				continue
+			}
+			key := CatalogKey{
+				Namespace: source.Namespace,
+				Name:      source.Name,
+			}
+			result[key] = &LazyClient{address: source.Address()}
+		}
+	}
+	return result
+}
+
+type CatalogDependencyCache interface {
+	GetCSVNameFromCatalog(csvName string, catalog CatalogKey) (Operator, error)
+	GetCSVNameFromAllCatalogs(csvName string) ([]Operator, error)
+	GetPackageFromAllCatalogs(pkg string) ([]Operator, error)
+	GetPackageVersionFromAllCatalogs(pkg string, version semver.Version) ([]Operator, error)
+}
+
+type OperatorCache struct {
+	logger    logrus.FieldLogger
+	rcp       RegistryClientProvider
+	snapshots map[CatalogKey]*CatalogSnapshot
+	ttl       time.Duration
+	sem       chan struct{}
+	m         sync.RWMutex
+}
+
+func NewOperatorCache(rcp RegistryClientProvider) *OperatorCache {
+	return &OperatorCache{
+		logger:    logrus.New(),
+		rcp:       rcp,
+		snapshots: make(map[CatalogKey]*CatalogSnapshot),
+		ttl:       5 * time.Minute,
+		sem:       make(chan struct{}, 4),
+	}
+}
+
+type NamespacedOperatorCache struct {
+	snapshots map[CatalogKey]*CatalogSnapshot
+}
+
+var _ CatalogDependencyCache = &NamespacedOperatorCache{}
+
+func (c *OperatorCache) Namespaced(namespaces ...string) *NamespacedOperatorCache {
+	now := time.Now()
+	clients := c.rcp.ClientsForNamespaces(namespaces...)
+
+	result := NamespacedOperatorCache{
+		snapshots: make(map[CatalogKey]*CatalogSnapshot),
+	}
+
+	var misses []CatalogKey
+	c.m.RLock()
```

Comment:
nit: I personally like using a nested function context for this. makes it harder to mess up in refactoring.

```go
func() {
  lock
  defer unlock
  doStuff 
}()
```